torch/fx/passes/reinplace.py

```diff
@@ -0,0 +1,540 @@
+            tree_map(_add_to_map, n.meta['fake_result'])
+
+    # inplace-ify functional ops, subject to the constraints written below.
+    all_later_view_inverse_node_usages = []
+    for idx, node in enumerate(gm.graph.nodes):
+        if node.op == 'call_function':
+            # Step 1: Check to see if this operator has an inplace variant.
+            maybe_inplace_op = _maybe_get_inplace_op(node.target)
+            if maybe_inplace_op is not None:
+                # This is a proxy check for ensuring that the first argument is "tensor-like"
+                # (This should be the case for all ops with inplace variants in ATen,
+                # although we technically don't have guarantees for custom ops).
+                assert len(node.target._schema.arguments) > 0
+                assert 'Tensor' in str(node.target._schema.arguments[0].type)
+
+                # Step 2: ensure that the op we're trying to re-inplace isn't a program input.
+                self_arg = node.args[0]
+                self_arg_name = self_arg.name
+                self_arg_storage = StorageWeakRef(self_arg.meta['fake_result'].storage())
+                if self_arg_storage in input_storages:
+                    # TODO: later, add the optimization for handling `copy_()` calls in the graph.
+                    continue
+
+                # Technically, grabbing every node in the graph with the same FakeTensor storage
+                # will give you all aliases of a node, plus other nodes like "return output".
+                # In this case, I want "self_aliases" to refer *only* to the nodes corresponding to actual tensors.
+                # In general, this is satisfied by aliasing nodes with targets that are torch operators.
+                self_aliases = storage_to_nodes[StorageWeakRef(self_arg.meta['fake_result'].storage())]
+                self_aliases = [
+                    x for x in self_aliases
+                    # Sigh, we also need to check for getitem(), which FX includes when grabbing tensors
+                    # from a multi-output view node.
+                    if isinstance(x.target, torch._ops.OpOverload) or x.target == _operator.getitem
+                ]
+
+                later_view_inverse_node_usages = _get_all_later_node_usages(self_arg, self_aliases, node.meta['node_idx'], only_include_view_inverse_nodes=True)
+                later_node_usages = _get_all_later_node_usages(self_arg, self_aliases, node.meta['node_idx'])
+
+                # Step 3: Check to see if the input to the op is re-used later in the graph.
+                # If not (same goes for its aliases), then this op is safe to re-in place.
+                # This is a slightly roundabout way to check that there are no later usages of the current self argument.
+                # (later_view_inverse_node_usages corresponds to "view_scatter" nodes that we are allowed to delete)
+                if len(later_node_usages - later_view_inverse_node_usages) == 0:
```

Comment:
Is it ever the case that `later_view_inverse_node_usages - later_node_usages` is nonzero